File: Linked Lists/2. Add Two Numbers/2. Add Two Numbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function (l1, l2) {
  const dummy = new ListNode();
  let curr = dummy;
  let carry = 0;

  while (l1 || l2) {
    const curr1 = l1 ? l1.val : 0;
    const curr2 = l2 ? l2.val : 0;

    let summ = curr1 + curr2 + carry;
    carry = Math.floor(summ / 10);
    summ %= 10;

    curr.next = new ListNode(summ);
    curr = curr.next;

    l1 = l1 ? l1.next : null;
    l2 = l2 ? l2.next : null;
  }

  if (carry) curr.next = new ListNode(carry);
  return dummy.next;
};

// Time: O(n)
// Space: O(1)

File: Linked Lists/2. Add Two Numbers/2. Add Two Numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        curr = dummy = ListNode()
        carry = 0

        while l1 or l2:
            curr1 = l1.val if l1 else 0
            curr2 = l2.val if l2 else 0
    
            summ = curr1 + curr2 + carry
            carry = summ // 10
            summ %= 10

            curr.next = ListNode(summ)
            curr = curr.next
            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None

        if carry:
            curr.next = ListNode(carry)
        
        return dummy.next

# Time: O(n)
# Space: O(1)